compute_tools/src/compute.rs

```diff
@@ -133,6 +133,84 @@ impl TryFrom<ComputeSpec> for ParsedSpec {
     }
 }
 
+/// Create special neon_superuser role, that's a slightly nerfed version of a real superuser
+/// that we give to customers
+fn create_neon_superuser(spec: &ComputeSpec, client: &mut Client) -> Result<()> {
+    let roles = spec
+        .cluster
+        .roles
+        .iter()
+        .map(|r| format!("'{}'", escape_literal(&r.name)))
+        .collect::<Vec<_>>();
+
+    let dbs = spec
+        .cluster
+        .databases
+        .iter()
+        .map(|db| format!("'{}'", escape_literal(&db.name)))
+        .collect::<Vec<_>>();
+
+    let roles_decl = if roles.is_empty() {
+        String::from("roles text[] := NULL;")
+    } else {
+        format!(
+            r#"
+               roles text[] := ARRAY(SELECT rolname
+                                     FROM pg_catalog.pg_roles
+                                     WHERE rolname IN ({}));"#,
+            roles.join(", ")
+        )
+    };
+
+    let database_decl = if dbs.is_empty() {
+        String::from("dbs text[] := NULL;")
+    } else {
+        format!(
+            r#"
+               dbs text[] := ARRAY(SELECT datname
+                                   FROM pg_catalog.pg_database
+                                   WHERE datname IN ({}));"#,
+            dbs.join(", ")
+        )
+    };
+
+    // ALL PRIVILEGES grants CREATE, CONNECT, and TEMPORARY on all databases
+    // (see https://www.postgresql.org/docs/current/ddl-priv.html)
+    let query = format!(
+        r#"
+            DO $$
+                DECLARE
+                    r text;
+                    {}
+                    {}
```

Comment:
The escaping isn't quite right here. Try creating a role or database with `$$` in its name.

Reply:
Hmm, if I'm reading this right, the above queries in `pg_roles` and `pg_databases` are always executed, even if `neon_superuser` role already exists. I think we should optimize for the case that `neon_superuser` already exists, and check for that first. The slow-path, when `neon_superuser` does not exist, is not so performance critical, so I think this construction of arrays and sending them all to Postgres in one statement is a premature optimization.

I'd suggest moving the control logic to Rust. It'll be more readable, and then you don't need to deal with the `$$` escaping of the plpgsql block. Sending everything to the server in one plpgsql block is faster, but I doubt the difference is that large when the client and server are on the same host, and the slow-path isn't very performance critical anyway.